File: backend/dosier_infrastructure/Curriculum/NormativaService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using dosier_application.Curriculum.Dtos;
using dosier_application.Curriculum.Interfaces;
using dosier_infrastructure.data.models;

namespace dosier_infrastructure.Curriculum
{
    public class NormativaService : INormativaService
    {
        private readonly DosierContext _context;

        public NormativaService(DosierContext context)
        {
            _context = context;
        }

        public async Task<List<NormativaDto>> GetNormativasVigentesAsync(string? organismo = null)
        {
            var query = _context.DocNormativas
                .AsNoTracking()
                .Include(n => n.Articulos.OrderBy(a => a.Orden))
                .Where(n => n.Activo);

            if (!string.IsNullOrWhiteSpace(organismo))
            {
                query = query.Where(n => n.OrganismoEmisor == organismo);
            }

            var list = await query.OrderBy(n => n.OrganismoEmisor).ThenBy(n => n.CodigoResolucion).ToListAsync();

            return list.Select(n => new NormativaDto
            {
                IdNormativa = n.IdNormativa,
                Uuid = n.Uuid,
                OrganismoEmisor = n.OrganismoEmisor,
                TipoNormativa = n.TipoNormativa,
                CodigoResolucion = n.CodigoResolucion,
                Titulo = n.Titulo,
                Descripcion = n.Descripcion,
                FechaEmision = n.FechaEmision,
                FechaVigencia = n.FechaVigencia,
                ArchivoUrl = n.ArchivoUrl,
                Activo = n.Activo,
                Articulos = n.Articulos.Select(a => new NormativaArticuloDto
                {
                    IdArticulo = a.IdArticulo,
                    Uuid = a.Uuid,
                    IdNormativa = a.IdNormativa,
                    CodigoResolucion = n.CodigoResolucion,
                    OrganismoEmisor = n.OrganismoEmisor,
                    NumeroArticulo = a.NumeroArticulo,
                    Titulo = a.Titulo,
                    Contenido = a.Contenido,
                    RequisitoCurricular = a.RequisitoCurricular,
                    Orden = a.Orden
                }).ToList()
            }).ToList();
        }

        public async Task<List<NormativaArticuloDto>> GetChecklistCurricularAsync(string? organismo = null)
        {
            var query = _context.DocNormativaArticulos
                .AsNoTracking()
                .Include(a => a.Normativa)
                .Where(a => a.Normativa != null && a.Normativa.Activo);

            if (!string.IsNullOrWhiteSpace(organismo))
            {
                query = query.Where(a => a.Normativa!.OrganismoEmisor == organismo);
            }

            var articulos = await query.OrderBy(a => a.Normativa!.OrganismoEmisor)
                                       .ThenBy(a => a.Orden)
                                       .ToListAsync();

            return articulos.Select(a => new NormativaArticuloDto
            {
                IdArticulo = a.IdArticulo,
                Uuid = a.Uuid,
                IdNormativa = a.IdNormativa,
                CodigoResolucion = a.Normativa?.CodigoResolucion ?? string.Empty,
                OrganismoEmisor = a.Normativa?.OrganismoEmisor ?? string.Empty,
                NumeroArticulo = a.NumeroArticulo,
                Titulo = a.Titulo,
                Contenido = a.Contenido,
                RequisitoCurricular = a.RequisitoCurricular,
                Orden = a.Orden
            }).ToList();
        }

        public async Task<ModeloEducativoDto?> GetModeloEducativoVigenteAsync()
        {
            var modelo = await _context.DocModelosEducativos
                .AsNoTracking()
                .Where(m => m.Activo)
                .OrderByDescending(m => m.FechaVigenciaDesde)
                .FirstOrDefaultAsync();

            if (modelo == null) return null;

            return new ModeloEducativoDto
            {
                IdModelo = modelo.IdModelo,
                Uuid = modelo.Uuid,
                Codigo = modelo.Codigo,
                Nombre = modelo.Nombre,
                Version = modelo.Version,
                ResolucionAprobacion = modelo.ResolucionAprobacion,
                Descripcion = modelo.Descripcion,
                FechaVigenciaDesde = modelo.FechaVigenciaDesde,
                FechaVigenciaHasta = modelo.FechaVigenciaHasta,
                ArchivoUrl = modelo.ArchivoUrl,
                Activo = modelo.Activo
            };
        }
    }
}

File: dosier_web/src/services/normativaService.ts
import api from '../api/axios_config';

// ─────────────────────────────────────────────────────────────
//  DTOs de Normativas Externas y Marco Regulatorio (CES / CACES)
// ─────────────────────────────────────────────────────────────

export interface NormativaArticuloDto {
    idArticulo: number;
    uuid: string;
    idNormativa: number;
    codigoResolucion: string;
    organismoEmisor: string;
    numeroArticulo: string;
    titulo?: string;
    contenido: string;
    requisitoCurricular?: string;
    orden: number;
}

export interface NormativaDto {
    idNormativa: number;
    uuid: string;
    organismoEmisor: 'CES' | 'CACES' | 'SENESCYT' | 'MINEDUC' | 'OTRO';
    tipoNormativa: string;
    codigoResolucion: string;
    titulo: string;
    descripcion?: string;
    fechaEmision?: string;
    fechaVigencia?: string;
    archivoUrl?: string;
    activo: boolean;
    articulos: NormativaArticuloDto[];
}

export interface ModeloEducativoDto {
    idModelo: number;
    uuid: string;
    codigo: string;
    nombre: string;
    version: string;
    resolucionAprobacion?: string;
    descripcion?: string;
    fechaVigenciaDesde: string;
    fechaVigenciaHasta?: string;
    archivoUrl?: string;
    activo: boolean;
}

export interface PerfilEgresoResultadoDto {
    idResultadoPerfil: number;
    uuid: string;
    idPerfilEgreso: number;
    codigo: string;
    descripcion: string;
    orden: number;
}

export interface PerfilEgresoDto {
    idPerfilEgreso: number;
    uuid: string;
    idCarrera: number;
    nombreCarrera?: string;
    idMalla: number;
    version: string;
    descripcionGeneral: string;
    activo: boolean;
    resultados: PerfilEgresoResultadoDto[];
}

export interface AsignaturaTributacionDto {
    idRelacion: number;
    idAsignatura: number;
    idMalla: number;
    idResultadoPerfil: number;
    codigoRda: string;
    descripcionRda: string;
    nivelAporte: 'Introductorio' | 'Medio' | 'Avanzado';
}

// ─────────────────────────────────────────────────────────────
//  Funciones de Cliente API (Axios con normalización snake_case)
// ─────────────────────────────────────────────────────────────

const mapArticulo = (raw: any): NormativaArticuloDto => ({
    idArticulo: raw.id_articulo ?? raw.idArticulo ?? 0,
    uuid: raw.uuid ?? '',
    idNormativa: raw.id_normativa ?? raw.idNormativa ?? 0,
    codigoResolucion: raw.codigo_resolucion ?? raw.codigoResolucion ?? '',
    organismoEmisor: raw.organismo_emisor ?? raw.organismoEmisor ?? '',
    numeroArticulo: raw.numero_articulo ?? raw.numeroArticulo ?? '',
    titulo: raw.titulo ?? '',
    contenido: raw.contenido ?? '',
    requisitoCurricular: raw.requisito_curricular ?? raw.requisitoCurricular ?? '',
    orden: raw.orden ?? 0
});

const mapNormativa = (raw: any): NormativaDto => ({
    idNormativa: raw.id_normativa ?? raw.idNormativa ?? 0,
    uuid: raw.uuid ?? '',
    organismoEmisor: raw.organismo_emisor ?? raw.organismoEmisor ?? 'CACES',
    tipoNormativa: raw.tipo_normativa ?? raw.tipoNormativa ?? '',
    codigoResolucion: raw.codigo_resolucion ?? raw.codigoResolucion ?? '',
    titulo: raw.titulo ?? '',
    descripcion: raw.descripcion ?? '',
    fechaEmision: raw.fecha_emision ?? raw.fechaEmision,
    fechaVigencia: raw.fecha_vigencia ?? raw.fechaVigencia,
    archivoUrl: raw.archivo_url ?? raw.archivoUrl,
    activo: Boolean(raw.activo),
    articulos: Array.isArray(raw.articulos) ? raw.articulos.map(mapArticulo) : []
});

export const getNormativas = async (organismo?: string): Promise<NormativaDto[]> => {
    const res = await api.get('/normativas', { params: organismo ? { organismo } : undefined });
    return Array.isArray(res.data) ? res.data.map(mapNormativa) : [];
};

export const getChecklistCurricular = async (organismo?: string): Promise<NormativaArticuloDto[]> => {
    const res = await api.get('/normativas/checklist', { params: organismo ? { organismo } : undefined });
    return Array.isArray(res.data) ? res.data.map(mapArticulo) : [];
};

export const getModeloEducativo = async (): Promise<ModeloEducativoDto | null> => {
    try {
        const res = await api.get('/normativas/modelo-educativo');
        const raw = res.data;
        if (!raw) return null;
        return {
            idModelo: raw.id_modelo ?? raw.idModelo ?? 0,
            uuid: raw.uuid ?? '',
            codigo: raw.codigo ?? '',
            nombre: raw.nombre ?? '',
            version: raw.version ?? '1.0',
            resolucionAprobacion: raw.resolucion_aprobacion ?? raw.resolucionAprobacion,
            descripcion: raw.descripcion,
            fechaVigenciaDesde: raw.fecha_vigencia_desde ?? raw.fechaVigenciaDesde,
            fechaVigenciaHasta: raw.fecha_vigencia_hasta ?? raw.fechaVigenciaHasta,
            archivoUrl: raw.archivo_url ?? raw.archivoUrl,
            activo: Boolean(raw.activo)
        };
    } catch {
        return null;
    }
};

export const getPerfilEgreso = async (idCarrera: number, idMalla: number): Promise<PerfilEgresoDto | null> => {
    try {
        const res = await api.get('/normativas/perfil-egreso', { params: { idCarrera, idMalla } });
        const raw = res.data;
        if (!raw) return null;
        return {
            idPerfilEgreso: raw.id_perfil_egreso ?? raw.idPerfilEgreso ?? 0,
            uuid: raw.uuid ?? '',
            idCarrera: raw.id_carrera ?? raw.idCarrera ?? 0,
            nombreCarrera: raw.nombre_carrera ?? raw.nombreCarrera,
            idMalla: raw.id_malla ?? raw.idMalla ?? 0,
            version: raw.version ?? '1.0',
            descripcionGeneral: raw.descripcion_general ?? raw.descripcionGeneral ?? '',
            activo: Boolean(raw.activo),
            resultados: Array.isArray(raw.resultados)
                ? raw.resultados.map((r: any) => ({
                      idResultadoPerfil: r.id_resultado_perfil ?? r.idResultadoPerfil ?? 0,
                      uuid: r.uuid ?? '',
                      idPerfilEgreso: r.id_perfil_egreso ?? r.idPerfilEgreso ?? 0,
                      codigo: r.codigo ?? '',
                      descripcion: r.descripcion ?? '',
                      orden: r.orden ?? 0
                  }))
                : []
        };
    } catch {
        return null;
    }
};

export const getTributacionAsignatura = async (idAsignatura: number, idMalla: number): Promise<AsignaturaTributacionDto[]> => {
    try {
        const res = await api.get('/normativas/tributacion-asignatura', { params: { idAsignatura, idMalla } });
        if (!Array.isArray(res.data)) return [];
        return res.data.map((r: any) => ({
            idRelacion: r.id_relacion ?? r.idRelacion ?? 0,
            idAsignatura: r.id_asignatura ?? r.idAsignatura ?? 0,
            idMalla: r.id_malla ?? r.idMalla ?? 0,
            idResultadoPerfil: r.id_resultado_perfil ?? r.idResultadoPerfil ?? 0,
            codigoRda: r.codigo_rda ?? r.codigoRda ?? '',
            descripcionRda: r.descripcion_rda ?? r.descripcionRda ?? '',
            nivelAporte: r.nivel_aporte ?? r.nivelAporte ?? 'Medio'
        }));
    } catch {
        return [];
    }
};
